### PR TITLE
Add io component

### PR DESCRIPTION
**Pull request by LHT129**
issue #40 
- implement memory block io
- io test for read&write
- io test for serialize&deserialize
- add common index param
- add datatype

### CODE REVIEW

File: src/io/basic_io_test.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+template <typename T>
+void
+TestSerializeAndDeserialize(BasicIO<T>& wio, BasicIO<T>& rio) {
+    std::vector<uint64_t> counts = {200, 500};
+    std::vector<uint64_t> max_lengths = {2, 20, 37, 64, 128, 260, 999, 4097};
+    srandom(time(nullptr));
+    std::string dirname = "/tmp/random_TestSerializeAndDeserialize_" + std::to_string(random());
+    std::filesystem::create_directory(dirname);
```
Comment (inabao) [inline src/io/basic_io_test.h]: use `fixtures::temp_dir` to create tmp directory for testing

File: src/io/memory_block_io.h [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+    inline void
+    SerializeImpl(StreamWriter& writer);
+
+    inline void
+    DeserializeImpl(StreamReader& reader);
+
+private:
+    [[nodiscard]] inline bool
+    checkValidOffset(uint64_t size) const {
```
Comment (inabao) [inline src/io/memory_block_io.h]: `checkValidOffset` to `check_valid_offset`

File: src/io/memory_block_io.h [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        return size <= blocks_.size() * block_size_;
+    }
+
+    inline void
+    checkAndRealloc(uint64_t size);
+
+    [[nodiscard]] inline const uint8_t*
+    getDataPtr(uint64_t offset) const {
```
Comment (inabao) [inline src/io/memory_block_io.h]: ditto

File: src/io/memory_block_io.h [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    DeserializeImpl(StreamReader& reader);
+
+private:
+    [[nodiscard]] inline bool
+    checkValidOffset(uint64_t size) const {
+        return size <= blocks_.size() * block_size_;
+    }
+
+    inline void
+    checkAndRealloc(uint64_t size);
```
Comment (inabao) [inline src/io/memory_block_io.h]: ditto

File: src/io/memory_block_io.h [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    [[nodiscard]] inline const uint8_t*
+    getDataPtr(uint64_t offset) const {
+        auto block_no = offset / block_size_;
+        auto block_off = offset % block_size_;
+        return blocks_[block_no] + block_off;
+    }
+
+    [[nodiscard]] inline bool
+    checkInOneBlock(uint64_t off1, uint64_t off2) const {
```
Comment (inabao) [inline src/io/memory_block_io.h]: ditto

File: src/io/memory_block_io.h [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        return (off1 / block_size_) == (off2 / block_size_);
+    }
+
+private:
+    uint64_t block_size_{DEFAULT_BLOCK_SIZE};
+
+    std::vector<uint8_t*> blocks_{};
+
+    Allocator* const allocator_{nullptr};
```
Comment (inabao) [inline src/io/memory_block_io.h:105]: `Allocator* const` to `const Allocator*`

File: tests/fixtures/fixtures.cpp [hunk 7]
```diff
@@ -200,4 +201,25 @@ brute_force(const vsag::DatasetPtr& query,
...
     return std::move(result);
 }
+
+std::vector<IOItem>
+GenTestItems(uint64_t count, uint64_t max_length, uint64_t max_index) {
+    std::vector<IOItem> result(count);
+    std::unordered_set<uint64_t> maps;
+    for (auto& item : result) {
+        while (true) {
+            item.start_ = (random() % max_index) * max_length;
```
Comment (ShawnShawnYou) [inline tests/fixtures/fixtures.cpp:211]: the real max_length of IO data is max_length^2 ?

File: src/io/memory_io.h [hunk 8]
```diff
@@ -68,7 +82,7 @@ class MemoryIO : public BasicIO<MemoryIO> {
     }
 
 private:
-    Allocator* allocator_{nullptr};
+    Allocator* const allocator_{nullptr};
```
Comment (inabao) [inline src/io/memory_io.h:85]: ditto

File: src/io/memory_block_io.h (hunk 6 above)
Comment (LHT129) [reply src/io/memory_block_io.h:105]: I think T* const is better

File: src/io/memory_io.h (hunk 8 above)
Comment (LHT129) [reply src/io/memory_io.h:85]: I think T* const is better

File: src/io/memory_block_io.h (hunk 5 above)
Comment (LHT129) [reply src/io/memory_block_io.h]: done

File: src/io/memory_block_io.h (hunk 3 above)
Comment (LHT129) [reply src/io/memory_block_io.h]: done

File: src/io/memory_block_io.h (hunk 4 above)
Comment (LHT129) [reply src/io/memory_block_io.h]: done

File: src/io/memory_block_io.h (hunk 2 above)
Comment (LHT129) [reply src/io/memory_block_io.h]: done

File: src/io/basic_io_test.h (hunk 1 above)
Comment (LHT129) [reply src/io/basic_io_test.h]: done

**Review by inabao — APPROVED** [review]

**Review by ShawnShawnYou — APPROVED** [review]

**Pull request merged**